src/components/ui/TagInput.tsx

```diff
@@ -0,0 +1,180 @@
+import { useState, KeyboardEvent, useEffect, forwardRef } from "react";
+import { Input } from "./input";
+import { X } from "lucide-react";
+import { cn } from "@/lib/utils";
+
+/*
+==== Sample Usage without react-hook-form ====
+
+const MyForm = () => {
+  const [tags, setTags] = useState(["tag1", "tag2"]);
+
+  const handleSubmit = (e: FormEvent) => {
+    e.preventDefault();
+    console.log(tags);
+  };
+
+  return (
+    <form onSubmit={handleSubmit}>
+      <TagInput name="tags" defaultValue={tags} onChange={setTags} />
+      <button type="submit">Submit</button>
+    </form>
+  );
+};
+
+
+==== Example with react-hook-form ====
+
+import { useForm, Controller } from "react-hook-form";
+
+interface FormData {
+  tags: string[];
+}
+
+const MyForm = () => {
+  const { control, handleSubmit } = useForm<FormData>();
+
+  const onSubmit = (data: FormData) => {
+    console.log(JSON.stringify(data, null, 2));
+  };
+
+  return (
+    <form onSubmit={handleSubmit(onSubmit)}>
+      <Controller
+        name="tags"
+        control={control}
+        defaultValue={[]}
+        render={({ field: { onChange, value, ref } }) => (
+          <TagInput
+            setValue={onChange}
+            name="tags"
+            defaultValue={value}
+            onChange={onChange}
+            ref={ref}
+          />
+        )}
+      />
+      <button type="submit">Submit</button>
+    </form>
+  );
+};
+
+
+==== Styling ====
+  list
+    tag
+  input  
+
+
+=== Note ===
+This Component disallows the repetition of tags
+*/
+
+interface TagInputProps {
+  setValue?: (name: string, value: string[]) => void;
+  name: string;
+  defaultValue?: string[];
+  onChange?: (tags: string[]) => void;
+  placeholder?: string;
+  listStyle?: string;
+  tagStyle?: string;
+  inputStyle?: string;
+}
+
+const TagInput = forwardRef<HTMLInputElement, TagInputProps>(
+  (
+    {
+      setValue,
+      name,
+      defaultValue = [],
+      onChange,
+      listStyle,
+      tagStyle,
+      inputStyle,
+      placeholder,
+    },
+    ref
+  ) => {
+    const [tags, setTags] = useState<string[]>(
+      Array.isArray(defaultValue) ? defaultValue : []
+    );
+
+    const [inputValue, setInputValue] = useState<string>("");
+
+    useEffect(() => {
+      // Only update tags if defaultValue changes
```

Comment:
Seems to be updating the tags if its default values get changed, or do you see a problem with it?
Though I don't see why the use effect is also monitoring the tags, since it only needs to update the tags to the new value if defaultValue gets changed.